### PR TITLE
project ready for marking

### PR DESCRIPTION
**Pull request by Am0031**
HTML and CSS were provided and left unchanged.

Work done:
added the code in javascript file for generating the password
checked functionality worked correctly for all scenarios
refactored code where possible for a cleaner code

Deployed page: https://am0031.github.io/password-generator
Github repository: https://github.com/Am0031/password-generator

### CODE REVIEW

File: assets/js/index.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+//function - get the password length from the user
+const getPasswordLength = () => {
+  //prompt user to give a userInput
+  //declare variable to receive user input - const -> not modified further down
+  const userInput = prompt(
+    "How long would you like your password to be? \n\nPlease enter a number between 8 and 128.\n",
+    ""
+  );
+  // check if userInput is a number
+  const passwordLength = parseInt(userInput);
```
Comment (surajverma2587) [inline assets/js/index.js]: Always good to pass the second radix argument when using the `parseInt` function which is `10`. See documentation for more information on that.

```suggestion
  const passwordLength = parseInt(userInput, 10);
```

File: assets/js/index.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    //if type and range are both true, the number is returned for storage in variable
+    return passwordLength;
+  }
+  //if type or range are not true, it sends an alert box with message
+  else {
+    alert(
+      "You must enter a number between 8 and 128. \n\nClick OK to start again."
+    );
+    //click ok on alert box to reload the page and start again
+    window.location.reload(true);
```
Comment (surajverma2587) [inline assets/js/index.js]: The boolean argument you passed is no longer supported and it seems like you might have referenced an old stack overflow article. Please always cross check with the official documentation after finding something online.

```suggestion
    window.location.reload();
```

File: assets/js/index.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    document.getElementById("password").reset();
```
Comment (surajverma2587) [inline assets/js/index.js]: This is not required as the `reset` method is used to reset form controls/inputs. The password field is not a form input and there is no form in the HTML layout. 

```suggestion
    
```

Comment (surajverma2587) [inline assets/js/index.js]: This line also throws a JS error in the console. Removing this feature should prevent that bug

`TypeError: document.getElementById(...).reset is not a function`

File: assets/js/index.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ",
+    "0123456789",
+    " !\"#$%&'()*+,-./:;<=>?@[]^_`{|}~",
+  ];
+  //declare variable for result array of strings of characters - start with empty array - const -> only pushing in it, not modifying
+  const choiceArray = [];
+
+  //prompt the user with a question and get true/false response
+  //4 prompts - loop for 4 questions
+  for (let i = 0; i < question.length; i += 1) {
```
Comment (surajverma2587) [inline assets/js/index.js]: Good implementation of using arrays for questions and characters. I would have opted for an array of objects instead where you have the question as a key and the characters as another key. This way the order of the `question` array is not dependent on the order of the `characters` array.

File: assets/js/index.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    window.location.reload(true);
+    document.getElementById("password").reset();
+  }
+};
+
+//function - get the user preferences on which characters to include in password
+const getPasswordOptions = () => {
+  //step 1: getUserOption part
+  //declare variables for questions and strings of characters
+  const question = [
```
Comment (surajverma2587) [inline assets/js/index.js]: Please rename this to `questions` as it indicates that the variable is a potential array

```suggestion
  const questions = [
```

File: assets/js/index.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    "0123456789",
+    " !\"#$%&'()*+,-./:;<=>?@[]^_`{|}~",
+  ];
+  //declare variable for result array of strings of characters - start with empty array - const -> only pushing in it, not modifying
+  const choiceArray = [];
+
+  //prompt the user with a question and get true/false response
+  //4 prompts - loop for 4 questions
+  for (let i = 0; i < question.length; i += 1) {
+    const getUserOption = confirm(question[i]); //const -> used only in this loop, not modified in the if statement
```
Comment (surajverma2587) [inline assets/js/index.js]: I would have named this as `userOption`. The prefix of `get` suggests that something is a function which will return something.

```suggestion
    const userOption = confirm(question[i]); //const -> used only in this loop, not modified in the if statement
```

File: assets/js/index.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  }
+  //step 2: isValidOption part
+  //check the options match the acceptance criteria => at least 1 type of characters selected
+  //if array.length = 0, then alert the user
+  if (choiceArray.length == 0) {
+    alert(
+      "You must choose at least one type of characters to include in your password. \n\nClick OK to start again."
+    );
+    //click ok on alert box to reload the page and start again
+    window.location.reload(true);
```
Comment (surajverma2587) [inline assets/js/index.js]: Same as comment above

File: assets/js/index.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    document.getElementById("password").reset();
```
Comment (surajverma2587) [inline assets/js/index.js]: Same as comment above

File: assets/js/index.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    );
+    //pushing character in variable
+    draftPassword.push(chosenOptions[chosenArray][randomNumber]);
+  }
+  //returns the newly formed collection of characters (as an array of strings of 1 character each)
+  return draftPassword;
+};
+
+//function createPassword 2/2 - shuffle the temporary password and convert it to a string
+const shufflePassword = (tempPassword) => {
```
Comment (surajverma2587) [inline assets/js/index.js:98]: Excellent use of this function to shuffle the temporary password

File: assets/js/index.js (hunk 3 above)
Comment (Am0031) [reply assets/js/index.js]: hi, when I remove that line, it removes the console error, but it also doesn't refresh the page, it actually just continues showing the prompts for the next questions after I close the password length alert. How can I fix that and actually find a way to stop after the password length when not good?

Comment (Am0031) [reply assets/js/index.js]: never mind, I resolved it! you can have a look and let me know what you think if you like :)

**Comment by Am0031** [comment]
thanks for the comments! I made the corrections and found a way to reload the page without throwing an error in the console. I hope that looks better :)

**Pull request merged**